libp2p/services/autonatservice.nim

```diff
@@ -0,0 +1,128 @@
+    autonat: Autonat
+    newStatusHandler: NewStatusHandler
+    rng: ref HmacDrbgContext
+    numPeersToAsk: int
+    maxConfidence: int
+
+
+  NetworkReachability* {.pure.} = enum
+    NotReachable, Reachable, Unknown
+
+  NewStatusHandler* = proc (networkReachability: NetworkReachability): Future[void]  {.gcsafe, raises: [Defect].}
+
+proc new*(
+  T: typedesc[AutonatService],
+  autonat: Autonat,
+  rng: ref HmacDrbgContext,
+  scheduleInterval: Option[Duration] = none(Duration),
+  numPeersToAsk: int = 5,
+  maxConfidence: int = 3): T =
+  return T(
+    scheduleInterval: scheduleInterval,
+    networkReachability: NetworkReachability.Unknown,
+    t: initCountTable[NetworkReachability](),
+    autonat: autonat,
+    rng: rng,
+    numPeersToAsk: numPeersToAsk,
+    maxConfidence: maxConfidence)
+
+proc networkReachability*(self: AutonatService): NetworkReachability {.inline.} =
+  return self.networkReachability
+
+proc handleAnswer(self: AutonatService, ans: NetworkReachability) {.async.} =
+  if ans == NetworkReachability.Unknown:
+    return
+  if ans == self.networkReachability:
+    if self.t[ans] == self.maxConfidence:
+      return
+    self.t.inc(ans)
+  else:
+    if self.t[self.networkReachability] > 0:
+      self.t.inc(self.networkReachability, -1)
+    if self.t[ans] < self.maxConfidence:
+      self.t.inc(ans)
+    if self.t[ans] == self.maxConfidence or self.t[self.networkReachability] == 0:
+      self.networkReachability = ans
+
+  if self.t[self.networkReachability] == self.maxConfidence:
+    if not isNil(self.newStatusHandler):
+      await self.newStatusHandler(self.networkReachability)
+
+  trace "Current status confidence", confidence = $self.t
+  trace "Current status", currentStats = $self.networkReachability
+
+proc askPeer(self: AutonatService, s: Switch, peerId: PeerId): Future[bool] {.async.} =
+  trace "Asking for reachability", peerId = $peerId
+  let (ans, known) =
+    try:
+      let ma = await self.autonat.dialMe(peerId)
+      (NetworkReachability.Reachable, true)
+    except AutonatUnreachableError:
+      (NetworkReachability.NotReachable, true)
+    except AutonatError:
+      (NetworkReachability.Unknown, false)
+  await self.handleAnswer(ans)
+  return known
+
+proc askConnectedPeers(self: AutonatService, switch: Switch) {.async.} =
+  var peers = switch.connectedPeers(Direction.Out)
+  self.rng.shuffle(peers)
+  var peersToAsk = min(self.numPeersToAsk, peers.len)
+  for peer in peers:
+    if peersToAsk == 0:
+      break
+    if await askPeer(self, switch, peer):
+      peersToAsk -= 1
+
+proc register(service: AutonatService, switch: Switch, interval: Duration) {.async.} =
+  heartbeat "Register AutonatService run", interval:
+    await service.run(switch)
+
+method setup*(self: AutonatService, switch: Switch): Future[bool] {.async.} =
+  let hasBeenSettedUp = await procCall Service(self).setup(switch)
+  if hasBeenSettedUp and self.scheduleInterval.isSome:
```

Comment:
should we force a `scheduleInterval` to be mandatory?

Reply:
Most likely.